### ISSUE TITLE
Upcoming Additions

### ISSUE THREAD

**Issue by shaps80** [shaps80/SwiftUIBackports#9]
In order to keep everyone up-to-date with what I'm working on and to avoid duplication, here's a list of the currently planned additions/backports:

- [ ] ImageRenderer
- [ ] Dismiss (environment)
- [ ] ShareLink
- [ ] DynamicTypeSize
- [ ] OpenURL modifier
- [ ] PasteButton
- [ ] QuickLook (modifier)
- [ ] ScrollDismissesKeyboard (modifier)
- [ ] ScrollIndicators (modifier)
- [ ] Searchable
- [ ] TextEditor
- [ ] Tint (modifier)
- [ ] LabelledContent
- [ ] NavigationStack

> Note, not all of the above will actually end up being back ported, as occasionally an implementation proves difficult or impossible to reproduce reliably.
> Also the list above is in no particular order.

**Comment by schiewe** [comment]
Hi @shaps80, great work! Are there any plans to backport the iOS 16 `PhotosPicker`?

**Comment by shaps80** [comment]
Potentially, I've done a similar thing in the past, including property wrappers that make fetching way easier and thus far simpler to build custom UI rapidly. Feel free to check it out: https://github.com/SwiftUI-Plus/Media

**Comment by pblondin** [comment]
Hi @shaps80, thanks a lot for your work! This is really helpful!

I'm curious, do you think it would be possible to backport the @[AccessibilityFocusState](https://developer.apple.com/documentation/swiftui/accessibilityfocusstate) wrapper? I honestly have not clue if it's even possible.

Thanks in advance! 🙏

**Comment by shaps80** [comment]
To be honest, I have no clue, and I don't have the time at the moment to investigate. Feel free to look into it yourself and comment here with more details and then I can have a deeper discussion about it. I just don't have the time myself at the moment, sorry

**Comment by tazmancoder** [comment]
Hi @shaps80, thanks for putting in all the work on this. I have a question regarding the new ShareLink. I currently use ActivitiyView, but I noticed in my fork there's a message at the top with the following:

> This backport is now available as Backport.ShareLink, in a single Backports library, with a LOT more additions. This should simply my efforts and allow me and others to contribute more backports in the near future. [SwiftUI Backports](https://github.com/shaps80/SwiftUIBackports)

But when I added the back ports library ShareLink wasn't there, when will this be there?
Also, will you be adding extra functionality like ActivityView has where there's an onComplete. I need this because I update my CoreData Store when the share is complete.

**Comment by shaps80** [comment]
Hey! Thanks for reaching out. Yeah sorry about the left in comment. I had to pull that because it didn't work close enough to expectation. I had planned to revisit the implementation but hadn't found the time atm. 

Regarding additional features I've only done this in a few key areas where the API was relevant. This would be a little more difficult to define given it's not supposed to be used in that way (at least currently).

That being said when I revisit this I'll have a think about those things. Tbh it'd be nice to support custom activity's etc too. 

But perhaps my SwiftUI+ lib would be better suited for this. Let me see what I can come up with 👍

**Comment by tazmancoder** [comment]
Hey,

Thanks for the reply I appreciate it. I will just have to continue using ActivityView since it does give me that functionality. I hope that ActivityView will be sticking around?

ShareLink, being new to us, maybe next year we’ll get that functionality. Of course if they don’t, I’ll have to come up with another way to know that something was shared.

Thanks again,

-Mark/*

Mark Perryman
Apple (MAC) Enthusiast
(435) 882-1008

**Comment by shaps80** [comment]
Yep ActivityView is def sticking around since its super useful. I’ll take another crack at ShareLink at some point, will keep this Issue open a little longer so I can reply to you if I manage something earlier 👍

**Comment by shaps80** [comment]
@tazmancoder great news, I found the old branch and realised I'd gone a lot further than I'd expected. I've filled in the blanks and it's pretty close now. Including iOS __and__ macOS support.

I had to include something _like_ `Transferable` but its obviously not a complete backport.

However there's a new API in the `CoreTransferable` that allows you to get progress and feedback when an `NSItemProvider` has finished loading some resource. I'm thinking this could be a nice approach for knowing when 'sharing' has completed, that's not directly connected to `ShareLink` which is a UI element.

This would likely suit your needs as well, since updating your model should occur once the resource has actually been shared, not when some UI has been dismissed.

Will keep you updated, but should have a branch, including this ShareLink implementation, by tomorrow 👍

**Comment by shaps80** [comment]
Its in the DemoApp (will be), has Apple's original header docs and literally matching API so look at any official demos from anyone/anywhere and you'll know how to use it, but its honestly super simple. Regardless Backports are all fully documented with demos:
[SwiftUIBackports Demo](https://github.com/shaps80/SwiftUIBackportsDemo)

**Comment by shaps80** [comment]
FYI: https://developer.apple.com/documentation/photokit/photospickeritem/4030160-loadtransferable?changes=__9

This isn't the API I'd end up including but in case its useful for your own reference.

**Comment by shaps80** [comment]
FYI I’ve fallen sick today, will get to this very soon, just need a couple days 👍🏻

**Comment by shaps80** [comment]
@tazmancoder ok so if you point to `main` you can try out the initial `ShareLink` implementation. It currently support `String`, `URL`, `Image` and `UI/NS Image` types.

In order to support custom previews and things like import/export including knowledge about transfers, I plan to also backport `Tranferable`. More specifically this will bring support for arbitrary `Data` and `Codable` types, making it far more useful without having to do the copying to file (for example) manually.

This will take a while longer but the current implementation is pretty usable for the most use-cases.

**Comment by tazmancoder** [comment]
Good morning,

I have a question, I am using ActivityView in one of my Apps. I have noticed that if you bring up the activity view and choose to email or text or anything else. If you cancel the email or text the activity view doesn’t dismiss. I can’t seem to figure out why this is not dismissing. Any help or advice would be appreciated….

Thanks,

-Mark/*

Mark J Perryman
Apple Product Enthusiast
(435) 882-1008

**Comment by shaps80** [comment]
@tazmancoder as mentioned above, this package does not use that implementation so I hope you don't think I'm ignoring the question but I don't want to flood this ticket which was honestly just for my own tracking of work in progress, in a transparent way.

Thanks.

**Comment by shaps80** [comment]
@schiewe not sure if I ever remembered to let you know, but a while back I did indeed backport the PhotosPicker API 👍🏻
